File: util/sqs.js
var AWS = require('aws-sdk'); 
var snsSqsConfig = require('../config/sns-sqs.js')[ process.env.STAGE || 'local' ];
var cacheUtil   = require('./cacheUtil');

// configure AWS
AWS.config.update({
  'region': snsSqsConfig.REGION
});

var sqs = new AWS.SQS();
var followService;

var receiveMessageParams = {
  QueueUrl: snsSqsConfig.QUEUE,
  MaxNumberOfMessages: 10
};



function getMessages() {
	console.log("reading msgs");
  sqs.receiveMessage(receiveMessageParams, receiveMessageCallback);
}

function receiveMessageCallback (err, data) {
	var message;
	var temp;
	var receiptHandle;
  if (data && data.Messages && data.Messages.length > 0) {
	    for (var i=0; i < data.Messages.length; i++) {
		      temp =  JSON.parse(data.Messages[i].Body);
		      message = JSON.parse(temp.Message);
		      receiptHandle = data.Messages[i].ReceiptHandle;
		      processMessage(message,receiptHandle); 
	    }
  } 
  
  if (err) {
	  console.log('Error while reading messages',err);
  }
}

function processMessage(message,receiptHandle) {
	console.log("Recivied the message, need to act on this..!!!",message, message.event, receiptHandle);
    if (message.event == 'USER.DELETE') {
    	var accessTokens = message.message.accessToken; 
    	for (var i=0; i < accessTokens.length; i++) {
    		cacheUtil.del(accessTokens[i]);
    	}
    	deleteMessage(receiptHandle);
    }
}

function deleteMessage(receiptHandle, index) {
	console.log("Deleteing msg with handle",receiptHandle);
	var deleteMessageParams = {
	    QueueUrl: snsSqsConfig.QUEUE
	};
	deleteMessageParams.ReceiptHandle = receiptHandle
	sqs.deleteMessage(deleteMessageParams, function(err, data) {
		if (err) {
			console.log('Error while deleting message',err);
		} 
			
	});
}


function SQS() {

}

SQS.prototype.init = function() {
	setInterval(getMessages, 15000);
}

module.exports = SQS;

File: src/utils/SQS.go
package utils

import (
	"time"
	"log"
	"auth/src/config"
	"os"
	"encoding/json"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/sqs"
)

type SqsMessage struct {
	Message string
}

type Message struct {
	Version string
	Meta Meta
	Name string
	Event string
	Message InnerMessage
}

type Meta struct {
	ResourceType string
	ResourceID int64
	ServiceID string
	ServiceVersion string
}

type InnerMessage struct {
	AccessToken []string
}

var svc *sqs.SQS

func SQSInit() {
	sess, err := session.NewSession(&aws.Config{
		Region: aws.String(config.SQS.Region)},
	)

	if err != nil {
		os.Exit(2)
	}

	svc = sqs.New(sess)
	receiveMessages()
}
func receiveMessages() {
	ticker := time.NewTicker(time.Duration(config.SQS.PollIntervalSeconds) * time.Second)
	log.Println(config.SQS.QueueURL)
	for _ = range ticker.C {
		result, err := svc.ReceiveMessage(&sqs.ReceiveMessageInput{
			AttributeNames: []*string{
				aws.String(sqs.MessageSystemAttributeNameSentTimestamp),
			},
			MessageAttributeNames: []*string{
				aws.String(sqs.QueueAttributeNameAll),
			},
			QueueUrl:            &config.SQS.QueueURL,
			MaxNumberOfMessages: aws.Int64(10),
			//VisibilityTimeout:   aws.Int64(36000),  // 10 hours
			//WaitTimeSeconds:     aws.Int64(0),
		})
		if (err != nil) {
			log.Println("Error while reading from queue ",err)
			ticker.Stop()
		}

		processMessages(result.Messages)
	}
}


func processMessages(sqsMessages []*sqs.Message) {
	for _,sqsMsg := range sqsMessages {
		var sqsMessage SqsMessage
		if err := json.Unmarshal([]byte(*sqsMsg.Body),&sqsMessage); err != nil {
			log.Println("Error while unmarshaling error ",err)
		}
		var message Message
		if err := json.Unmarshal([]byte(sqsMessage.Message),&message); err != nil {
			log.Println("Error while unmarshalling 2 error", err)
		}

		if message.Version == "2.0" && message.Event == "USER.DELETE" {
		//log.Println(message.Message.AccessToken)
			if len(message.Message.AccessToken) > 0 {
				for _,at := range message.Message.AccessToken {
					DeleteCache(at)
				}
			}
		}
		deleteMessage(sqsMsg.ReceiptHandle)
	}
}


func deleteMessage(receiptHandle *string) {
	_, err := svc.DeleteMessage(&sqs.DeleteMessageInput{
		QueueUrl:      &config.SQS.QueueURL,
		ReceiptHandle: receiptHandle,
	})

	if err != nil {
		log.Println("Delete Error", err)
	}

}
